File: problem1.go
// Time Complexity :  O(N)
// Space Complexity : O(1) 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
func productExceptSelf(nums []int) []int {
    if len(nums) == 0{
        return []int{}
    }
    
    rp:=1
    
    result := make([]int,len(nums))
    result[0] = rp
    //left running product
    for i:=1;i<len(nums);i++{
        rp = rp*nums[i-1]
        result[i] = rp
    }

    rp = 1
    result[len(nums)-1] =  result[len(nums)-1]*rp
    for i:=len(nums)-2;i>=0;i--{
        rp = rp*nums[i+1]
        result[i] = result[i]*rp
    }
    
    return result
    
}

File: Problem1.cs
public class Solution
    {
        // Time Complexity : O(n)
        // Space Complexity : O(1)
        // Did this code successfully run on Leetcode : Yes
        // Any problem you faced while coding this : No
        public int[] ProductExceptSelf(int[] nums)
        {
            int n = nums.Length;
            int runningProduct = 1;
            int[] result = new int[n];

            result[0] = 1;

            //Calculate the product of all elements to the left of each element
            for(int i = 1; i < n; i++)
            {
                runningProduct = runningProduct * nums[i - 1];
                result[i] = runningProduct;
            }

            runningProduct = 1;//reset

            //calculate the product of all elements to the right of each element
            for (int i = n - 2; i >= 0; i--)
            {
                runningProduct = runningProduct * nums[i + 1];
                result[i] = result[i] * runningProduct; //since we are saving the right product in the same result array multiple with the left product
            }

            return result;

        }
    }
